geniza/corpus/management/commands/sync_annotation_export.py

```diff
@@ -1,22 +1,132 @@
+
+            # TODO: handle deletions (object no longer in db)
+
+            # make a dictionary of modified annotations and the users who modified them
+            modified_annotations = defaultdict(list)
+            for log_entry in log_entries:
+                modified_annotations[log_entry.object_id].append(log_entry.user)
+
+            # load the modified annotations from the database
+            annotations = Annotation.objects.filter(
+                id__in=list(modified_annotations.keys())
+            )
+            # group by manifest, so we can export by document
+            annos_by_manifest = annotations.group_by_manifest()
+
+            for manifest, annotations in annos_by_manifest.items():
+                # export transcription for the specified document,
+                # documenting the users who modified it
+                document = Document.from_manifest_uri(manifest)
+
+                # collect all users who modified any of the annotations
+                # for this document based on the collected log entries
+                users = set()
+                for anno in annotations:
+                    # NOTE: annotation id is a uuid; must cast to string
+                    # for dict lookup to succeeed
+                    users |= set(modified_annotations[str(anno.id)])
+
+                self.anno_exporter.export(
+                    pgpids=[document.pk],
+                    modifying_users=users,
+                )
+
+            # push changes to remote
+            self.anno_exporter.sync_github()
+
+        # update the last run for the next time
+        self.update_lastrun_info(new_lastrun)
+
+    def get_lastrun_info(self):
+        # check for information about the last run of this script;
+        # load as json if it is exists
+        if os.path.exists(self.lastrun_filename):
+            with open(self.lastrun_filename) as lastrun:
+                # load and parse as json
+                return json.load(lastrun)
+
+    def update_lastrun_info(self, new_lastrun):
+        # Update or create last run information file
+        lastrun_info = self.get_lastrun_info() or {}
+        lastrun_info.update({self.script_id: new_lastrun.isoformat()})
+        with open(self.lastrun_filename, "w") as lastrun:
+            return json.dump(lastrun_info, lastrun, indent=2)
+
+    def script_lastrun(self):
+        # determine the datetime for the last run of this script
+
+        # load information about the last run of this script
+        lastrun_data = self.get_lastrun_info()
+        # if the file exists, pull out modified value for this scriptdi
+        if lastrun_data and self.script_id in lastrun_data:
+            return datetime.fromisoformat(lastrun_data[self.script_id])
+
+        # if lastrun file is not found, use last git commit on the repository
+        # (potentially unreliable if non-data repo content is updated
+        # and lastrun file does not exist, but should be ok.)
```

Comment:
Could potentially mitigate with protected branches if this is a big concern, though sounds like it's not. (Also from a quick glance, not entirely sure there's a protected branch rule we could use here in the first place…)

Reply:
Another possibility would be some kind of identifier in the commit message that indicates the commit was done programmatically by the exporter, and then finding the last git commit that matches this identifier.

Reply:
Oh, interesting idea. I _think_ in normal operation we should have the last run file in place, so my hope is that the last git commit is just a reasonable, logical fallback so we don't re-export too much history. But you're right, we could probably filter fairly easily on commit messages if we need to.